src/probe_modules/module_tcp_synscan.c

```diff
@@ -83,71 +88,111 @@ void synscan_print_packet(FILE *fp, void *packet)
+			if (htonl(tcp->th_ack) != htonl(validation[0]) &&
+			    htonl(tcp->th_ack) != htonl(validation[0]) + 1) {
+				return PACKET_INVALID;
+			}
+		} else {
+			// For non RST packets, recv(ack) == sent(seq) + 1
+			if (htonl(tcp->th_ack) != htonl(validation[0]) + 1) {
+				return PACKET_INVALID;
+			}
+		}
+	} else if (ip_hdr->ip_p == IPPROTO_ICMP) {
+		struct ip *ip_inner;
+		size_t ip_inner_len;
+		if (icmp_helper_validate(ip_hdr, len, sizeof(struct tcphdr),
+					 &ip_inner,
+					 &ip_inner_len) == PACKET_INVALID) {
+			return PACKET_INVALID;
+		}
+		struct tcphdr *tcp = get_tcp_header(ip_inner, ip_inner_len);
+		if (!tcp) {
+			return PACKET_INVALID;
+		}
+		// we can always check the destination port because this is the
+		// original packet and wouldn't have been altered by something
+		// responding on a different port
+		port_h_t sport = ntohs(tcp->th_sport);
+		port_h_t dport = ntohs(tcp->th_dport);
+		if (dport != target_port) {
+			return PACKET_INVALID;
+		}
+		validate_gen(ip_hdr->ip_dst.s_addr, ip_inner->ip_dst.s_addr,
+			     (uint8_t *)validation);
+		if (!check_dst_port(sport, num_ports, validation)) {
+			return PACKET_INVALID;
+		}
+	} else {
+		return PACKET_INVALID;
 	}
-
-	return 1;
+	return PACKET_VALID;
 }
 
 static void synscan_process_packet(const u_char *packet,
-				   __attribute__((unused)) uint32_t len,
+				   UNUSED uint32_t len,
 				   fieldset_t *fs,
-				   __attribute__((unused)) uint32_t *validation)
+				   UNUSED uint32_t *validation)
 {
-	struct ip *ip_hdr = (struct ip *)&packet[sizeof(struct ether_header)];
-	struct tcphdr *tcp =
-	    (struct tcphdr *)((char *)ip_hdr + 4 * ip_hdr->ip_hl);
-
-	fs_add_uint64(fs, "sport", (uint64_t)ntohs(tcp->th_sport));
-	fs_add_uint64(fs, "dport", (uint64_t)ntohs(tcp->th_dport));
-	fs_add_uint64(fs, "seqnum", (uint64_t)ntohl(tcp->th_seq));
-	fs_add_uint64(fs, "acknum", (uint64_t)ntohl(tcp->th_ack));
-	fs_add_uint64(fs, "window", (uint64_t)ntohs(tcp->th_win));
-
-	if (tcp->th_flags & TH_RST) { // RST packet
-		fs_add_string(fs, "classification", (char *)"rst", 0);
+	struct ip *ip_hdr = get_ip_header(packet, len);
+	assert(ip_hdr);
+	if (ip_hdr->ip_p == IPPROTO_TCP) {
+		struct tcphdr *tcp = get_tcp_header(ip_hdr, len);
+		assert(tcp);
+		fs_add_uint64(fs, "sport", (uint64_t)ntohs(tcp->th_sport));
+		fs_add_uint64(fs, "dport", (uint64_t)ntohs(tcp->th_dport));
+		fs_add_uint64(fs, "seqnum", (uint64_t)ntohl(tcp->th_seq));
+		fs_add_uint64(fs, "acknum", (uint64_t)ntohl(tcp->th_ack));
+		fs_add_uint64(fs, "window", (uint64_t)ntohs(tcp->th_win));
+		if (tcp->th_flags & TH_RST) { // RST packet
+			fs_add_string(fs, "classification", (char *)"rst", 0);
+			fs_add_bool(fs, "success", 0);
+		} else { // SYNACK packet
+			fs_add_string(fs, "classification", (char *)"synack",
+				      0);
+			fs_add_bool(fs, "success", 1);
+		}
+		fs_add_null_icmp(fs);
+	} else if (ip_hdr->ip_p == IPPROTO_ICMP) {
+		// tcp
+		fs_add_null(fs, "sport");
```

Comment:
Is adding sport, dport, and classification here **and** in fs_populate_icmp_from_iphdr intentional? If not, I'll fix that.